Scan stake pool registrations during wallet restoration.

jonathanknowles opened the pull request:
# Issue Number

#899 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

I have:

- [x] Provided an implementation of `IsOurs (SeqState n k) ChimericAccount`.
- [x] Adjusted the return type of `applyBlock` so that it references any stake delegation that has occurred on behalf of the specified wallet.
- [x] Adjusted the definition of `restoreBlocks` so that it correctly processes the registrations returned by `applyBlocks`.
- [x] Adjusted the definition of `applyBlock` to identify delegations relevant to the specified wallet, and return the final registration as part of `FilteredBlock`.
- [x] Added tests to ensure that `applyBlock` does successfully discover delegations made by the wallet under test, and only that wallet.
- [x]  Added tests to ensure that a wallet's delegation status is correctly updated in the context of discovered delegations.

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

jonathanknowles committed:
Fix documentation comments for functions.

Fix comment for function `putDelegationCertificate`.
Fix comment for function `applyBlock`.
Fix comment for function `prefilterBlock`.

jonathanknowles committed:
Use more idiomatic `where` clause in function `applyBlock`.

jonathanknowles committed:
Introduce record type `FilteredBlock`.

Use this type to replace the list of transactions returned by
`applyBlock` and `applyBlocks`.

Future changes will add more fields to this type.

jonathanknowles committed:
Add `delegations` field to type `FilteredBlock`.

This represents the set of stake delegations made on behalf of the
wallet, listed in order of discovery.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -269,20 +271,35 @@ updateState
-        )
+    :: Wallet s
+    -> Block
+    -> WalletStateUpdate s
```
Introducing a type helper for the result of this function makes sense with the addition of the delegation certificate but this type has _too many responsibility_ and obfuscate a very idiomatic Haskell pattern (`s -> (result, s)`). 
Furthermore, the `initialState` sounds superfluous, this is already given as argument and, because of immutability, is already in-scope and unchanged at call sites.

I'd highly recommend the following as a middle-ground, preserving the state pattern while still allowing for a more readable output:
```hs
:: Wallet s
-> Block 
-> (ApplyBlockResult, Wallet s)
```
where:

```hs
 data ApplyBlockResult = ApplyBlockResult
    { ourTransactions :: ![(Tx, TxMeta)]
    , ourCertificates :: ![(PoolId, ChimericAccount)] -- Or alternatively ![PoolId]
    }
```

Also, delegation should really be a list (e.g. `[(PoolId, ChimericAccount)]` or `[PoolId]`) since there can be many certificates present in a single block :+1:

jonathanknowles replied:
@KtorZ wrote:
> I'd highly recommend the following as a middle-ground, preserving the state pattern while still allowing for a more readable output:
> ```hs
> :: Wallet s
> -> Block 
> -> (ApplyBlockResult, Wallet s)
> ```
Fair enough! I've updated the types to reflect this recommendation.

As for:
> Also, **delegation should really be a list** (e.g. [(PoolId, ChimericAccount)] or [PoolId]) since there can be many certificates present in a single block.

I checked and you're right that a _single block_ can encode _more than one_ delegation of stake from one account to _different pools_.

However, given this is the case, which (if any) of these delegations should take precedence? 

Our database (currently) only has space for one pool per `(wallet id, slot id)` combination:

https://github.com/input-output-hk/cardano-wallet/blob/c2700b15a147ab8ec744d812ef483b9614597005/lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs#L146-L149

So if I've understood correctly, we'd either have to pick one from the list, or adjust the database in some way.

I've asked the Jörmungandr team to clarify this here: https://app.slack.com/client/T0N639Z4N/threads/thread/CCPKJ93Q9-1574305614.149500

jonathanknowles replied:
Further to previous comment.

The answer seems to be that the _**latest**_ delegation (in a block) for an account should take precedence over others. See: https://input-output-rnd.slack.com/archives/CCPKJ93Q9/p1574312569153400?thread_ts=1574305614.149500&cid=CCPKJ93Q9

KtorZ replied:
Of course :)
And that is what the database take care of already.

However, the internals of the primitive models shouldn't make such assumptions IMO and yield all certificates discovered in the block, in the same order as they're discovered. 

Yet that is debatable. One could argue that hiding this "detail" inside the primitive model would also make sense and prevent the result to be misinterpreted. I have no strong arguments that would favor one over the other.

jonathanknowles replied:
@KtorZ wrote:
> One could argue that hiding this "detail" inside the primitive model would also make sense and prevent the result to be misinterpreted. I have no strong arguments that would favor one over the other.

Assuming that `applyBlock` yields all delegation certificates discovered in a block, the caller will probably still need to filter out all but the last.

jonathanknowles replied:
> I have no strong arguments that would favor one over the other.

If the **same block** contains _multiple_ delegations for the _same wallet_ to _different pools_, I think there's a strong argument for writing this information to the log (rather than silently dropping every delegation except the last).

Since we have given `restoreBlocks` the responsibility of logging relevant data discovered in each block, I think it makes sense here to return _**all**_ delegations discovered in the `ApplyBlockResult` type. This way `restoreBlocks` can log each of them.

jonathanknowles committed:
Adjust code to comply with team coding standard.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -269,20 +271,35 @@ updateState
-        )
+    :: Wallet s
+    -> Block
+    -> (ApplyBlockResult, Wallet s)
```
**[suggestion ]** 

Change back to `Block -> Wallet s -> (ApplyBlockResult, Wallet s)`

**Why?**

As mentioned before, this is a rather common pattern in Haskell, `s -> (result, s)`. Having `s` as the last argument of the function makes it easier to re-use functions out of the box like [state](https://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-State-Lazy.html#v:state). 

See also:

- [genChange](https://input-output-hk.github.io/cardano-wallet/haddock/edge/cardano-wallet-core-2019.11.18/Cardano-Wallet-Primitive-AddressDiscovery.html#v:genChange)
- [isOurs](https://input-output-hk.github.io/cardano-wallet/haddock/edge/cardano-wallet-core-2019.11.18/Cardano-Wallet-Primitive-AddressDiscovery.html#v:isOurs)

jonathanknowles replied:
> Change back to `Block -> Wallet s -> (ApplyBlockResult, Wallet s)`

Done.

jonathanknowles committed:
Write per-slot pool delegation certificates to the database.

When restoring blocks.

Additionally, write a log entry for each certificate discovered.

jonathanknowles committed:
Adjust the type of `preFilterBlock` to return a `FilteredBlock`.

For the moment, we return an empty list for the list of delegations.

jonathanknowles committed:
Use `unzip` to simplify `prop_applyBlockTxHistoryIncoming`.

jonathanknowles committed:
Return real list of stake pool delegations from function `prefilterBlock`.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -583,6 +590,11 @@ restoreBlocks ctx wid blocks nodeTip = db & \DBLayer{..} -> do
+    mapExceptT atomically $
+        forM_ slotPoolDelegations $ \delegation@(slotId, poolId) -> do
+            liftIO $ logDelegation delegation
+            putDelegationCertificate (PrimaryKey wid) poolId slotId
```
:ok_hand:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -386,26 +401,32 @@ blockchainParameters (Wallet _ _ _ bp) = bp
 prefilterBlock b u0 = runState $ do
-    (ourTxs, ourU) <- foldM applyTx (mempty, u0) (transactions b)
-    return (ourTxs, ourU)
+    delegations <- fmap snd <$> filterM isOurDelegation (b ^. #delegations)
```
**[suggestion]**

Use [mapMaybeM](https://hackage.haskell.org/package/extra-1.6.18/docs/Control-Monad-Extra.html#v:mapMaybeM)

**Why**

Avoid traversing the list twice.

jonathanknowles replied:
@KtorZ wrote:
> **[suggestion]**
> Use [mapMaybeM](https://hackage.haskell.org/package/extra-1.6.18/docs/Control-Monad-Extra.html#v:mapMaybeM)

I've followed this suggestion. See:

 https://github.com/input-output-hk/cardano-wallet/compare/6ffbaebf325b9fd35caf69a86424894686a10992..96c00ac44b5973ebbf9b39129a5114c9f611b848

KtorZ reviewed (approved):
LGTM

jonathanknowles committed:
Provide implementation of `IsOurs (SeqState n k) ChimericAccount`.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -532,9 +533,11 @@ instance
+    isOurs (ChimericAccount account) state =
+        (account == ourAccount, state)
+      where
+        ourAccount = xpubPublicKey $ getRawKey $ rewardAccountKey state
```
:ok_hand:

KtorZ committed:
provide clear function to convert a reward account to a chimeric account

KtorZ committed:
finalize remaining integration tests about delegation

KtorZ committed:
fix race-condition in STAKE_POOLS tests
